teps/0026-interceptor-plugins.md

```diff
@@ -0,0 +1,510 @@
+}
+```
+
+With this, an example JSON payload sent to the CEL Interceptor will look like the following:
+```json
+{
+  "apiVersion": "triggers.tekton.dev/v1alpha1",
+  "kind": "InterceptorMessage",
+  "request": {
+    "body": {}, // some body
+    "header": {
+      "eventType": "push",
+    }, 
+    "interceptor_params": {
+      "filter": "header.Match(\"eventType\",\"push\"")
+    },
+    "context": {
+      "event_id": "random-string",
+      "trigger_id": "namespaces/default/triggers/cel-trigger"
+    }
+  }
+}
+```
+And a sucessful response back will be as follows:
+
+```json
+{
+  "apiVersion": "triggers.tekton.dev/v1alpha1",
+  "kind": "InterceptorMessage",
+  "response": {
+    "continue": true,
+  }
+}
+```
+While a failed response will look like:
+
+```json
+{
+  "apiVersion": "triggers.tekton.dev/v1alpha1",
+  "kind": "InterceptorMessage",
+  "response": {
+    "continue": false,
+    "status": {
+      "code": "400", //Failed Precondition?
+      "message": "filter evaluated to false"
+    },
+  }
+}
+```
+
+### Installing Interceptors
+Interceptors are installed via the InterceptorConfiguration CRD. At minimum
+the spec declares a `clientConfig` field that describes how the EventListener
+should send requests to the Interceptor Service.
+The interceptor spec can also contain option parameters fields that declares
+any parameters that an interceptor requires. Use cases for this include
+expressions for CEL interceptor, a reference to shared secret for
+GitHub/GitLab interceptors etc.
+
+The InterceptorConfiguration CRD conforms to the
+[Callable](https://github.com/knative/eventing/blob/master/docs/spec/interfaces.md#callable)
+Interface contract from Knative.
+
+In the future, we can consider explicitly allowing Interceptors to support
+one ore more version of the Interceptor API by adding a supportedVersions
+field to the `spec` (similar to [AdmissionReviewVersions](https://github.com/kubernetes/api/blob/master/admissionregistration/v1beta1/types.go#L452))
+
+```go
+type InterceptorConfigurationSpec struct {
+	// ClientConfig defines how to communicate with the interceptor service
+	// Required
+	ClientConfig InterceptorClientConfig `json:"clientConfig"`
+
+	// Params declare interceptor specific fields thatthe user can configure per Trigger.
+  Params []InterceptorParamSpec `json:"params"`
+  
+}
+
+type InterceptorClientConfig struct {
+	// `url` gives the location of the webhook, in standard URL form
+	// (`scheme://host:port/path`). Exactly one of `url` or `service`
+	// must be specified.
+	// +optional
+	URL *string `json:"url,omitempty"`
+
+	// `service` is a reference to the service for this webhook. Either
+	// `service` or `url` must be specified.
+	// +optional
+	Service *ServiceReference `json:"service,omitempty"`
+
+  // `caBundle` is a PEM encoded CA bundle which will be used to validate the webhook's server certificate.
```

Comment:
@savitaashture @bigkevmcd added an optional `caBundle` field here for TLS cert verification.